### ISSUE TITLE
[New Feature] extend CUDAPluggableAllocator to support caching algorithms of segments, blocks and calling frames

### ISSUE THREAD

**Issue by yiakwy-xpu-ml-framework-team** [pytorch/pytorch#115336]
### 🚀 The feature, motivation and pitch

The current implementation does not allow re-use pytorch memory pool caching strategy over external allocator:

```
void CUDAPluggableAllocator::cacheInfo(int dev_id, size_t* largestBlock) {
  TORCH_CHECK(
      false,
      "CUDAPluggableAllocator does not yet support cacheInfo. "
      "If you need it, please file an issue describing your use case.");
}
```

The **CUDAPluggableAllocator** just allocates what a person requests and no caching is applied :

<img width="650" alt="calling custom allocator" src="https://github.com/pytorch/pytorch/assets/89890040/9aa9b3f5-1938-49fc-8bbc-f4f97b908a84">

Hence for a new allocator, such as tracer allocator,  AMD unified memory allocator(#107605), we have to do memory management by ourselves :  pieces/segments, bins/blocks, and frames. 

Moreover, the **aten IR** has a very deep path to associate tensor, or op identities with the underlying allocation info (timeline, bytes, streamId, deviceId).

### Alternatives

The first solution involves inspecting aten IR such that memory timeline and liveness info can be directly attributed to aten IR and upper python calling frame.

The second solution still separates allocator and aten Ir but leaves a allocator an interface to interactively modify **AllocInfo**. This later should be added to python wrapper or hooks, so that use will not sense existence of this modification.

**Comment by youkaichao** [comment]
is it possible to make `CUDAPluggableAllocator` work under the caching allocator? i.e. we can reuse the caching behavior, and only replace the `cudaMalloc` function call from the caching allocator, when the memory is not enough.

**Comment by youkaichao** [comment]
described in https://dev-discuss.pytorch.org/t/understanding-the-difference-between-the-caching-behavior-of-cuda-caching-allocator-and-pluggable-allocator/2746/3 , it would be better if we can have a new argument like `enable_caching=True`, so that `CUDAPluggableAllocator` can also reuse the caching behavior of PyTorch.

**Comment by yiakwy-xpu-ml-framework-team** [comment]
> is it possible to make `CUDAPluggableAllocator` work under the caching allocator? i.e. we can reuse the caching behavior, and only replace the `cudaMalloc` function call from the caching allocator, when the memory is not enough.

Yes it is a feature desired and under way. It is cool to hear that you have some experiments in driver side to Intercept cumalloc functions.

Previously I believe this feature is better natively supported in Pytorch because recording memory usages cross streams is challenges outside of torch (hacking in driver (your example), CachedCUDAPluggableAllocator shared libary (implemented a year ago) for example).

As for the cached allocation algorithm itself used in pytorch, I made a picture last year :

![Image](https://github.com/user-attachments/assets/d7f52df0-6d67-4910-aad1-7465c9afc76e)

**Comment by youkaichao** [comment]
@yiakwy-xpu-ml-framework-team I added in https://dev-discuss.pytorch.org/t/understanding-the-difference-between-the-caching-behavior-of-cuda-caching-allocator-and-pluggable-allocator/2746/4?u=youkaichao , we can have some caching behavior when we use CUDAPluggableAllocator through memory pool, although it does not have the full caching behavior.

**Comment by yiakwy-xpu-ml-framework-team** [comment]
> [@yiakwy-xpu-ml-framework-team](https://github.com/yiakwy-xpu-ml-framework-team) I added in https://dev-discuss.pytorch.org/t/understanding-the-difference-between-the-caching-behavior-of-cuda-caching-allocator-and-pluggable-allocator/2746/4?u=youkaichao , we can have some caching behavior when we use CUDAPluggableAllocator through memory pool, although it does not have the full caching behavior.

Yeah, the trick part is multi stream event API in torch : streams rely on stream event API to know whether other stream is using the memory block.
